manubot/cite/citekey.py

```diff
@@ -309,3 +309,44 @@ def infer_citekey_prefix(citekey):
         if citekey.lower().startswith(prefix):
             return prefix + citekey[len(prefix) :]
     return f"raw:{citekey}"
+
+
+def url_to_citekey(url):
+    """
+    Convert a HTTP(s) URL into a citekey.
+    For supproted sources, convert from url citekey to an alternative source like doi.
+    """
+    from urllib.parse import urlparse, unquote
+
+    parsed_url = urlparse(url)
+    if parsed_url.hostname.split(".")[-2:] == ["doi", "org"]:
+        # DOI URLs
+        doi = unquote(parsed_url.path.lstrip("/"))
+        return f"doi:{doi}"
+    if parsed_url.hostname.split(".")[-2:] == ["biorxiv", "org"]:
+        # bioRxiv URL to DOI
+        (biorxiv_id,) = re.findall(r"/([0-9]{6})", parsed_url.path)
+        return f"doi:10.1101/{biorxiv_id}"
```

Comment:
@rabdill this is snippet to extract the DOI for a preprint from its biorxiv URL. I couldn't find much documentation on bioRxiv IDs and how the URL corresponds to the DOI. Do you know if the regex is correct? Are all IDs 6 digits? Would matching on 6+ digits be more future proof (will bioRxiv need to start issuing 7 digit IDs in future)?

Reply:
I haven't seen any documentation either. I'm assuming you're referring to the URI structure that looks like `"/content/early/2019/01/22/${id}"`, right? As far as I can tell, all the canonical URLs now use the new format, `/content/10.1101/${id}v1"`, where the DOI is basically the last thing in the URL before the version identifier. Does that simplify anything for you, or do you want to support both URL formats?

In any case, this regex matches all the DOIs that are currently indexed, but I'd guess it could jump to 7 digits at some point—I think the `10.1101` prefix is used by other stuff from Cold Spring Harbor Labs as well. There's also no reason I know of that they'd stay numeric, either. Would taking everything before the version indicator be too liberal? i.e. `r"(10.1101/\w+)v\d+$"`

Reply:
Thanks @rabdill for the insights, super helpful.

> Does that simplify anything for you, or do you want to support both URL formats?

Would like to support the following URLs from the tests (all of which were generated as of yesterday):

https://github.com/manubot/manubot/blob/d9f4676f485e1e0f09047437f8cf6d6d25cbe345/manubot/cite/tests/test_citekey.py#L134-L139

Note the PDF link still uses the archaic URL format. 

> There's also no reason I know of that they'd stay numeric, either. Would taking everything before the version indicator be too liberal?

I'm worried about the URLs where `10.1101/` does not prefix the identifier. Maybe for now I'll change the regex to match 6+ digits to allow for expansion of the identifier space. Can always update later if alphabetical characters are added to identifiers. The nature of this function is that it's okay to fail to extract the DOI, but its not okay to extract an incorrect DOI.

Reply:
Well, that didn't take long. In case you didn't catch this, looks like the DOI pattern just changed: https://twitter.com/cshperspectives/status/1205109455752454144

Reply:
So new URLs are like:

```
https://www.biorxiv.org/content/10.1101/2019.12.11.872580v1
https://www.biorxiv.org/content/10.1101/2019.12.11.872580v1.full.pdf+html
https://www.biorxiv.org/content/10.1101/2019.12.11.872580v1.full.pdf
```

Old URLs have not changed. The implementation would be simpler if we dropped support for legacy PDF links like `https://www.biorxiv.org/content/early/2017/08/31/087619.full.pdf`.